url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageARMv5.java

```diff
@@ -0,0 +1,4246 @@
+
+        switch (id) {
+        case InstructionARM.OPCODE_S_ADC:
+        case InstructionARM.OPCODE_S_ADD:
+        case InstructionARM.OPCODE_S_AND:
+        case InstructionARM.OPCODE_S_BIC:
+        case InstructionARM.OPCODE_S_EOR:
+        case InstructionARM.OPCODE_S_ORR:
+        case InstructionARM.OPCODE_S_RSB:
+        case InstructionARM.OPCODE_S_RSC:
+        case InstructionARM.OPCODE_S_SBC:
+        case InstructionARM.OPCODE_S_SUB:
+            //with S bit
+            strInst = String.format("%s%s%s", inst.getOpcodeFieldName(),
+                    inst.getCondFieldName(),
+                    (s) ? "s" : "");
+            //rd, rn, shifter_operand
+            strOperand = String.format("%s, %s, %s", getRegName(rd),
+                    getRegName(rn), getAddrMode1Name(inst));
+            break;
+        case InstructionARM.OPCODE_S_MOV:
+        case InstructionARM.OPCODE_S_MVN:
+            //with S bit
+            strInst = String.format("%s%s%s", inst.getOpcodeFieldName(),
+                    inst.getCondFieldName(),
+                    (s) ? "s" : "");
+            //rd, shifter_operand
+            strOperand = String.format("%s, %s", getRegName(rd),
+                    getAddrMode1Name(inst));
+            break;
+        case InstructionARM.OPCODE_S_CMN:
+        case InstructionARM.OPCODE_S_CMP:
+        case InstructionARM.OPCODE_S_TEQ:
+        case InstructionARM.OPCODE_S_TST:
+            //S bit is 1
+            strInst = String.format("%s%s", inst.getOpcodeFieldName(),
+                    inst.getCondFieldName());
+            //rn, shifter_operand
+            strOperand = String.format("%s, %s", getRegName(rn),
+                    getAddrMode1Name(inst));
+            break;
+        default:
+            throw new IllegalArgumentException("Unknown opcode S-bit ID " +
+                    String.format("%d.", id));
+        }
+        printDisasm(inst, strInst, strOperand);
+    }
+
+    /**
+     * 論理積命令。
+     *
+     * @param inst ARM 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executeALUAnd(InstructionARM inst, boolean exec) {
+        int id = inst.getOpcodeSBitShiftID();
+        boolean s = inst.getSBit();
+        int rn = inst.getRnField();
+        int rd = inst.getRdField();
+        int opr = getAddrMode1(inst);
+        int left, right, dest;
+
+        if (!exec) {
+            printALUDisasm(inst, id);
+            return;
+        }
+
+        if (!inst.satisfiesCond(getCPSR())) {
+            return;
+        }
+
+        left = getReg(rn);
+        right = opr;
+        dest = left & right;
+
+        if (s && rd == 15) {
+            getCPSR().setValue(getSPSR());
+        } else if (s) {
+            getCPSR().setNBit(BitOp.getBit32(dest, 31));
+            getCPSR().setZBit(dest == 0);
+            getCPSR().setCBit(getAddrMode1Carry(inst));
+            //V flag is unaffected
+        }
+
+        setReg(rd, dest);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1227:1256, 1266:1295, 1702:1731, 1784:1813.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.